File: app/src/main/java/com/github/nickpakhomov/hotel_viewer/handler/FetcherHandlerImpl.java
package com.github.nickpakhomov.hotel_viewer.handler;

import android.content.Context;
import android.widget.Toast;

import com.github.nickpakhomov.hotel_viewer.R;
import com.github.nickpakhomov.hotel_viewer.models.ResponseBody;
import com.github.nickpakhomov.hotel_viewer.service.ApiClient;
import com.github.nickpakhomov.hotel_viewer.service.HotelService;

import java.net.UnknownHostException;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Helper class to process HTTP requests along with "refresh" animation
 * Created by Nikolay Pakhomov on 28/08/17.
 */

public class FetcherHandlerImpl implements FetcherHandler {
    private static final String LOG_TAG = FetcherHandlerImpl.class.getSimpleName();
    private AnimationHandlerImpl mAnimationHandler;
    
    private int lastHotelID;
    
    private Context mContext;
    
    private CompositeDisposable mCompositeDisposable;
    private HotelService hotelService;
    private QueryHandler mQueryHandler;
    
    public FetcherHandlerImpl(Context context, AnimationHandlerImpl animationHandler) {
        mContext = context;
        mAnimationHandler = animationHandler;
        
        mCompositeDisposable = new CompositeDisposable();
        hotelService = ApiClient.getClient().create(HotelService.class);
        mQueryHandler = new QueryHandlerImpl(context.getContentResolver());
    }
    
    /**
     * Fetching hotel previews to form a list.
     */
    public void fetchHotelPreviewsData() {
        mAnimationHandler.startAnimation();
        
        mCompositeDisposable.add(hotelService.fetchHotelPreviewsData()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(this::handleResponsePreviews, this::handleError));
    }
    
    /**
     * Fetching detailed hotel
     *
     * @param hotelID hotel ID to fetch the info from
     */
    public void fetchHotelsData(int hotelID) {
        lastHotelID = hotelID;
        mAnimationHandler.startAnimation();
        
        mCompositeDisposable.add(hotelService.fetchHotelData(hotelID)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(this::handleResponseHotel, this::handleError));
    }
    
    /**
     * Callback class to handle Fetch Hotel Previews request.
     */
    private void handleResponsePreviews(List<ResponseBody> responseBodies) {
        mAnimationHandler.stopAnimation();
        mQueryHandler.insert(responseBodies);
    }
    
    /**
     * Callback class to handle Fetch Hotel Data request.
     */
    private void handleResponseHotel(ResponseBody responseBody) {
        mAnimationHandler.stopAnimation();
        mQueryHandler.update(responseBody);
    }
    
    public void handleError(Throwable error) {
        mAnimationHandler.setStatusError();
        if (error instanceof UnknownHostException)
            Toast.makeText(mContext, mContext.getString(R.string.error_internet_connection), Toast.LENGTH_SHORT).show();
    }
    
    /**
     * Returning last fetched hotel ID in order to update it from ActionBar menu_main
     */
    int getLastHotelID() {
        return lastHotelID;
    }
    
    public void onDestroy() {
        if (mCompositeDisposable != null) {
            mCompositeDisposable.clear();
            mCompositeDisposable = null;
        }
        mAnimationHandler = null;
        mContext = null;
        hotelService = null;
        mQueryHandler = null;
    }
}

File: app/src/main/java/com/github/nickpakhomov/hotel_viewer/handler/FetcherHandlerImpl.kt
package com.github.nickpakhomov.hotel_viewer.handler

import android.content.Context
import android.widget.Toast
import com.github.nickpakhomov.hotel_viewer.R
import com.github.nickpakhomov.hotel_viewer.models.HotelBody
import com.github.nickpakhomov.hotel_viewer.service.ApiClient
import com.github.nickpakhomov.hotel_viewer.service.HotelService
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import java.net.UnknownHostException

/**
 * Helper class to process HTTP requests along with "refresh" animation
 * Created by Nikolay Pakhomov on 28/08/17.
 */

class FetcherHandlerImpl(
        private var mContext: Context,
        private var mAnimationHandler: AnimationHandler) : FetcherHandler {

    /**
     * Returning last fetched hotel ID in order to update it from ActionBar menu_main
     */
    private var lastHotelID: Int = 0

    private var mCompositeDisposable: CompositeDisposable = CompositeDisposable()
    private var hotelService: HotelService = ApiClient.client.create(HotelService::class.java)
    private var mQueryHandler: QueryHandler = QueryHandlerImpl(mContext.contentResolver)

    /**
     * Fetching hotel previews to form a list.
     */
    override fun fetchHotelPreviewsData() {
        mAnimationHandler.startAnimation()

        mCompositeDisposable.add(hotelService.fetchHotelPreviewsData()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe({ this.handleResponsePreviews(it) }, { this.handleError(it) }))
    }

    /**
     * Fetching detailed hotel
     *
     * @param hotelID hotel ID to fetch the info from
     */
    override fun fetchHotelsData(hotelID: Int) {
        lastHotelID = hotelID
        mAnimationHandler.startAnimation()

        mCompositeDisposable.add(hotelService.fetchHotelData(hotelID)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe({ this.handleResponseHotel(it) }, { this.handleError(it) }))
    }

    /**
     * Callback class to handle Fetch Hotel Previews request.
     */
    private fun handleResponsePreviews(hotelBodies: List<HotelBody>) {
        mAnimationHandler.stopAnimation()
        mQueryHandler.insert(hotelBodies)
    }

    /**
     * Callback class to handle Fetch Hotel Data request.
     */
    private fun handleResponseHotel(hotelBody: HotelBody) {
        mAnimationHandler.stopAnimation()
        mQueryHandler.update(hotelBody)
    }

    override fun handleError(error: Throwable) {
        mAnimationHandler.setStatusError()
        if (error is UnknownHostException)
            Toast.makeText(mContext, mContext.getString(R.string.error_internet_connection), Toast.LENGTH_SHORT).show()
    }

    override fun getLastHotelID(): Int = lastHotelID

    override fun onDestroy() {
        mCompositeDisposable.clear()
    }

    companion object {
        private val LOG_TAG = FetcherHandlerImpl::class.java.simpleName
    }
}
